### ISSUE TITLE
Leverage transforms framework for continuous enrichment

### ISSUE THREAD

**Issue by bczifra** [elastic/elasticsearch#85228]
### Description

https://github.com/elastic/elasticsearch/pull/73407 is an attempt to introduce "instant enrich policies" that:
>don't need to be executed and the enrich processor performs the lookups directly against the enrich policy's source index.

This would allow a user to:
1. Update the enrichment data (without have to execute the enrichment each time)
2. Index a new document against a pipeline using an enrich processor using that enrichment data
3. Have that new document, post-enrichment, reflect the new enrichment data.

What it would _not_ do is:
1. Update an existing document with the new enrichment data
2. Enrich a document from multiple indices containing enrichment data

Would it be make sense to extend [data transforms](https://www.elastic.co/guide/en/elasticsearch/reference/current/transforms.html) to do something like the following example (similar to the [Example: Enrich your data based on exact values](https://www.elastic.co/guide/en/elasticsearch/reference/current/match-enrich-policy-type.html))?
```
PUT /user_meta/_doc/1?refresh=wait_for
{
  "email": "mardy.brown@asciidocsmith.com",
  "first_name": "Mardy",
  "last_name": "Brown",
  "city": "New Orleans",
  "county": "Orleans",
  "state": "LA",
  "zip": 70116,
}

PUT /user/_doc/1
{
  "email": "mardy.brown@asciidocsmith.com",
  "account_id": "123456"
}

POST _transform/_preview
{
  "source": {
    "index": ["user_meta","users"]
  },
  "dest" : { 
    "index" : "transformed"
  },
  "pivot": {
    "group_by": { 
      "user": { "terms": { "field": "email" }}
    },
    "fields": [
      "email",
      "first_name",
      "last_name"
      "city",
      "state",
      "zip",
      "account_id"
    ]
  }
}
```
The resulting transform document would look like:
```
{
      "email": "mardy.brown@asciidocsmith.com",
      "first_name": "Mardy",
      "last_name": "Brown",
      "city": "New Orleans",
      "state": "LA",
      "zip": 70116,
      "account_id": 123456,
}
```

If the `user_meta` index is updated, say with:
```
PUT /user_meta/_doc/1?refresh=wait_for
{
  "email": "mardy.brown@asciidocsmith.com",
  "first_name": "Mardy",
  "last_name": "Brown",
  "city": "Santa Monica",
  "county": "Los Angeles",
  "state": "CA",
  "zip": 90401
}
```
then the transform index will update the transformed document to reflect the change in city/count/state/zip (which in ES is a DELETE followed by a PUT). 

An arbitrary number of enrichment indices could be supported by specifying the `index` property as an array (which transforms already support, I believe).

**Comment by hendrikmuhs** [comment]
You can implement this with the current feature set of transform today:

First of all you need timestamps, e.g. ingest timestamps, to run in continuous mode, which is required for transform in order to react on an update (there are thoughts to run continuous mode without timestamps, but that's a different story).

Once you have a timestamp you can implement this using the `top_metrics` aggregation on the `fields`:

```
"meta": {
      "top_metrics": {
        "metrics": [
          {"field": "email"},
          {"field": "first_name"},
          {"field": "last_name"},
          {"field": "city"},
...
        ],
        "sort": {"ingest_timestamp": "desc"}
      }
    }
```

(The (ingest) timestamp might be useful for a "last updated" field.)

The downside of this is the nested output:

```
{
   "email": "mardy.brown@asciidocsmith.com",
   "meta": {
      "first_name": "Mardy",
      "last_name": "Brown",
      "city": "New Orleans",
      "state": "LA",
      "zip": 70116,
      "account_id": 123456
  }
}
```

You can move it top level in an ingest pipeline.

To sum it up: Your use case is possible today, however it isn't very user-friendly to implement. We have some discussions going on to improve certain logic to make the API easier to use. We will consider this issue as part of it.

**Comment by bczifra** [comment]
Thanks @hendrikmuhs , I didn't realize that [`top_metrics`](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-top-metrics.html) aggregation supported `keyword` fields, that's what does the trick for this example (and perhaps most use cases), though it is intrinsically limited by the types supported by `top_metrics`.
